Update IdPs/SPs from SAML metadata

James-REANNZ opened the pull request:
Some IdPs/SPs are added directly to the SAML service. The reporting service sources it's information from the federation registry, meaning it doesn't see those entities. This change pulls metadata from the SAML service to augment the data from the federation registry.

James-REANNZ committed:
new: Add `domain` column to `organisations`

This is necessary to identify organisations independently from the
federation registry ID. Populate it when syncing from the FR.

James-REANNZ committed:
new: add saml_update.rb

James-REANNZ committed:
fix: move fetching/parsing metadata outside of transaction

vladimir-mencl-eresearch commented on .rubocop.yml:
```diff
@@ -30,3 +30,6 @@ Rails/InverseOf:
 
+Metrics/ClassLength:
+  Exclude:
+    - app/jobs/update_from_saml_service.rb
```
In the AAF code base, I've generally seen per-class exclusions done inline, for from `.rubocop.yml`

vladimir-mencl-eresearch replied:
The `.rubocop.yml` file needs rebasing anyway, I'll change this in the rebase.

vladimir-mencl-eresearch commented on Gemfile.lock:
```diff
@@ -370,17 +370,20 @@ GEM
 DEPENDENCIES
   aaf-gumboot
-  aaf-lipstick (>= 4.5.1)
+  aaf-lipstick (>= 4.5.1)!
```
Why the exclamation mark at the end of the line?

vladimir-mencl-eresearch replied:
Got it in master in the meantime as well - I assume newer bundler syntax.

vladimir-mencl-eresearch commented:
Hi @James-REANNZ ,

Thanks for the PR and sorry about the long delay in getting back to you.  School holidays and then sick children got in the way.  (Plus when I saw you were already working out the thoughts for a replacement, I thought this might be low priority anyway....)

But here are my thoughts on the PR.

(1) I like how you approached this with parsing the metadata from the SAML service (and found code to reuse for dealing with the metadata parsing).

(2) I would like to see a clear position on what remains as the role of the original sync from the Federation Registry.  I see the `clean` method is modified to only purge attribute definitions that have not been seen by the sync.  But it appears it would still attempt to synchronise (copy in) everything it used to sync.

All IdPs and SPs seen in the FR would also be seen in SAML Service.  So it might be right to remove them from the scope of the sync.

A tricky bit is that Organizations would only be seen in the SAML Service metadata if they have at least one IdP or SP - while FR also exposes Organizations that are "empty shells".  But the only benefit of having these Organizations visible in the Reporting Service is for the report on Federation Growth (number of registrations) - which I do not see as this crucial.

Perhaps this is more a question for the new Proposal for Reporting Service replacement (whether to sync empty Organizations).

But my main point here is to see a clear position on how the new sync from SAML Service interacts with the existing sync from FR.

But, if the new proposal moves forward, all of this may be just academic discussion.

I have a bit more feedback on individual aspects and I'll try attaching that to specific lines in the code.

But this is the main summary: like the approach, but want to see the clear position on how the two mechanisms would interact.

And thanks for getting it into the PR!

Cheers,
Vlad

vladimir-mencl-eresearch commented on app/jobs/update_from_saml_service.rb:
```diff
@@ -0,0 +1,161 @@
+
+  def org_identifier(name)
+    # Generate a valid identifier for an organisation given the OrganisationName (domain)
+    digest = OpenSSL::Digest.new('SHA256').digest("tuakiri:subscriber:#{name}")
```
I'm a little concerned about having two `org_identifier` methods that provide the same function but implement it differently - potential source for confusion.
https://github.com/REANNZ/reporting-service/blob/master/app/jobs/concerns/query_federation_registry.rb#L13

vladimir-mencl-eresearch replied:
Renamed to avoid confusion.

vladimir-mencl-eresearch commented on app/jobs/update_from_saml_service.rb:
```diff
@@ -0,0 +1,161 @@
+  def process_idp(node, entity_id, org, reg_date)
+    idp = IdentityProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
```
Watch out - there might be multiple instances of `mdui:DisplayName`, distinguished by the `xml:lang` attribute.  You'd want `xml:lang="en"`.  In Tuakiri, you'd likely only see the `en` versions, but eduGAIN-imported metadata might have more languages.

vladimir-mencl-eresearch replied:
Sorted.

vladimir-mencl-eresearch commented on app/jobs/update_from_saml_service.rb:
```diff
@@ -0,0 +1,161 @@
+  def process_sp(node, entity_id, org, reg_date)
+    sp = ServiceProvider.find_or_initialize_by(entity_id: entity_id)
+
+    name_node = xpath_at(node, './md:Extensions/mdui:UIInfo/mdui:DisplayName')
```
Same comment - should expect multiple instances and pick the `en` version.

vladimir-mencl-eresearch replied:
Sorted

vladimir-mencl-eresearch commented on app/jobs/update_from_saml_service.rb:
```diff
@@ -0,0 +1,161 @@
+
+  def process_org(node)
+    org_id_node = xpath_at(node, './md:OrganizationName')
+    org_name_node = xpath_at(node, './md:OrganizationDisplayName')
```
Same comment - should expect multiple instances and pick the `en` version.

vladimir-mencl-eresearch replied:
Sorted

vladimir-mencl-eresearch commented on config/reporting_service.yml.dist:
```diff
@@ -15,6 +15,9 @@ federationregistry:
   secret: 'This is the shared secret used for authenticating to the Rapid export API'
+saml_service:
+  metadata_url: https://saml.test.url/mdq
+  # metadata_cert: "The Certificate"
```
It wasn't clear to me right away whether this was meant to be a path to the certificate or the certificate embedded directly.  Probably obvious (and just not to me), but maybe the sample value could make it even clearer?

vladimir-mencl-eresearch replied:
As  configuration is now via env vars, which is not exactly suitable for passing full certificates, I'll change the setting to use path to certificate instead.

vladimir-mencl-eresearch commented on db/migrate/20220413213421_add_domain_to_organisation.rb:
```diff
@@ -0,0 +1,6 @@
+class AddDomainToOrganisation < ActiveRecord::Migration[6.0]
```
I have seen AAF committing the magic string:
`# frozen_string_literal: true`
as the first line of all Ruby code, including migrations - I expect Rubocop will complain if it isn't there...

vladimir-mencl-eresearch replied:
No longer required by Rubocop.

vladimir-mencl-eresearch commented:
Done with the review. Happy to discuss offline (Slack or Zoom) - though not sure how much this is worth pursuing - also depends on how the Reporting Service Replacement proposal progresses.

James-REANNZ committed:
new: Pull information from SAML service

Uses the metadata published by the SAML service to populate the
reporting service.

Entities that are registered directly with the SAML service aren't
visible in the reporting service, which limits the utility of the
reports as it misses entities such as the Hosted IdP instances.

James-REANNZ committed:
fix: Delete old attribute associations when importing from SAML service

James-REANNZ committed:
fix: don't delete most objects when importing from FR

Stop it from deleting objects that might have been imported from the
SAML service.

James-REANNZ committed:
new: verify metadata against certificate

If the certificate is not available, it trusts the metadata as given.

James-REANNZ committed:
bau: Fix rubocop violations

Exclude the `UpdateFromSAMLService` class from the `Metrics/ClassLength`
Cop.

vladimir-mencl-eresearch committed:
fix: override rubocop Metrics/ClassLength limit for app/jobs/update_from_saml_service.rb

vladimir-mencl-eresearch committed:
fix: bin/saml_update.rb: use correct class name

vladimir-mencl-eresearch committed:
new: saml_metadata: expect cert as path to file

(because configuration is now via env vars - not exactly suitable for passing full certificates)

vladimir-mencl-eresearch committed:
new: rename module to UpdateFromSAMLMetadata

To decouple it from the SAML service - can consume any SAML metadata.

vladimir-mencl-eresearch committed:
new: add UpdateFromSAMLMetadata module configuration

Has been dropped in rebase as ReportingService no longer reads a configuration file.

Add to new style env-var-driven configuration in config/reporting_service_configuration.rb

vladimir-mencl-eresearch committed:
fix: saml_metadata: select "en" DisplayName

Organisations, IdPs and SP may have multiple names in different languages/locales.

Look for an "en" DisplayName, or at least one where the locale starts with "en" (such as "en_NZ")

vladimir-mencl-eresearch committed:
fix: saml_metadata: reset deactivated_at when reactivating

vladimir-mencl-eresearch committed:
fix: saml_metadata: rename org_identifier method to avoid clash

Rename org_identifier method to avoid clash with same-named method in UpdateFromFR.

Also rename org_id variable to org_domain to avoid confusion.

vladimir-mencl-eresearch committed:
new: saml_metadata: generate transparent (temporary) org identifiers

Organisations are still managed by Federation Registry as authoritative source.

To avoid clashes when an organisation gets created from metadata, mark them
with transparent temporary identifiers that can be later replaced with proper
identifiers created based on their fr_id values.

vladimir-mencl-eresearch committed:
fix: saml_metadata: update org data from metadata only before synced from FR

To avoid changing back-and-forth against FR.

vladimir-mencl-eresearch committed:
fix: UpdateFromFR: handle orgs created from SAML metadata

Find them by domain name and assign them a proper FR-based identifier.

vladimir-mencl-eresearch committed:
fix: UpdateFR: skip IdP+SP sync if SamlMetadata update is used

To avoid flapping IdPs and SPs from two different sources.
Specifically, an IdP (or SP) inactive in FR but active in metadata
(which is the case for IdPs migrated to Hosted IdP) would flap
activte/inactive state between the two syncs.

vladimir-mencl-eresearch committed:
fix: fr_update: only fix organisation IDs if SAML metadata sync is enabled

James-REANNZ commented on app/jobs/update_from_saml_metadata.rb:
```diff
@@ -0,0 +1,174 @@
+  # rubocop:disable Metrics/MethodLength, Metrics/AbcSize
+  def process_org(node)
+    org_domain_node = xpath_at(node, './md:OrganizationName')
+    org_name_node = xpath_at(node, "./md:OrganizationDisplayName[@xml:lang='en']") ||
```
Do we want to handle the case where there isn't an english name, but there is a non-english name? Or is that not worth handling?

vladimir-mencl-eresearch replied:
Thanks, good point.  I thought of it as an edge-case with a question mark.  But if it ever gets used to ingest eduGAIN metadata, anything should be expected - including no name at all.

Pushed now - thanks for the suggestion.

James-REANNZ reviewed (commented):
Looks good to me (and not just because half of it is my work :P).

vladimir-mencl-eresearch committed:
new: saml_metadata: use non-English names if "en" name not found - and entity_id as last resort

(OrganizationDisplayName does not need a last-resort option as it will always be present in an Organization element)
